### ISSUE TITLE
Caching Streamed styles only

### ISSUE THREAD

**Issue by BernardoMariano** [styled-components/styled-components#1712]
I have one page in my current stack, with the same CSS always, but the contents of this page may be different based on a query param `/?id=0123456789`

This way, if I cache the whole buffered HTML per URL `req.originalUrl`, I would need to cache **every possible resulted page** with query param, re-collecting all my styles for every case of `/?id=`.

There could be a way to get the generated css from streamed styles while styled-components is interleaving them, in order to cache it specifically.

```js
const renderStream = sheet.interleaveWithNodeStream(
  renderToNodeStream(jsx),
  // Introducing this new second argument, that is a callback called
  // whenever the styles of a streamed chunk of HTML are collected
  (styles) => {
    // And then we can cache our styles progressively wherever we want.
    // Literally `<style data-styled-components="fnsQwL eLAIgU..."></style>`
  }
)
```

I used to be able to do that with `renderToString` because it would **return** the styles as a string:

```js
const sheet = new ServerStyleSheet()
sheet.collectStyles(renderToString(jsx))
const styles = sheet.getStyleTags()
// Hooray! I'll never need to collect my styles anymore :D
```

Any thoughts?

**Comment by mxstbr** [comment]
Maybe this helps? https://zeit.co/blog/streaming-server-rendering-at-spectrum

**Comment by BernardoMariano** [comment]
I found two issues with this caching approach: it caches the whole final generated HTML document, not just the CSS and when there's a cache hit, you `res.send(html)` synchronously from the cache, not making use of the advantages of the async stream rendering.

I'm right now implementing what I think is the best solution for this, will open a PR soon

**Comment by mxstbr** [comment]
> you res.send(html) synchronously from the cache, not making use of the advantages of the async stream rendering.

True, that should be updated—PR very welcome! https://github.com/withspectrum/spectrum
